code/week7/main.cpp

```diff
@@ -0,0 +1,186 @@
+int getMax(int a, int b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+/*
+when getMax<double> the method would look like
+double getMax(double a, double b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+
+// you can have multiple templates in a function
+template <class Foo, class Bar>
+Bar getMin(Foo a, Bar b) {
+    if(a < b) {
+        return a;
+    } else {
+        return b;
+    }
+}
+
+
+
+int main() {
+    using namespace std;
+
+    /*************** LIST ***************/
+    // http://www.cplusplus.com/reference/list/list/
+    // list<TYPE> name
+    list<int> int_list = {1,2,3};
+    print_data(int_list.begin(), int_list.end(), "int_list");
+
+    cout << "the first element is = " << int_list.front() << endl;
+    cout << "the last element is = " << int_list.back() << endl;
+
+    /* pop removes an element and destructs it
+     * push adds an element
+     * front does the action to the front
+     * back does the action to the back
+     */
+    int_list.pop_front();
+    int_list.push_back(53);
+    int_list.push_front(15);
+    int_list.pop_back();
+    print_data(int_list.begin(), int_list.end(), "int_list");
+
+    /*************** QUEUE ***************/
+    // http://www.cplusplus.com/reference/queue/
+    queue<char> queue;
+    queue.push('r');
+    queue.push('o');
+    queue.push('b');
+    queue.push('o');
+    queue.push('j');
+
+    while(!queue.empty()) {
+        cout << queue.front();
+        queue.pop();
+    }
+
+    /*************** STACK ***************/
+    // http://www.cplusplus.com/reference/stack/stack/
+    stack<char> stack;
+    stack.push('s');
+    stack.push('t');
+    stack.push('e');
+    stack.push('k');
+    stack.push('c');
+    stack.push('a');
+
+    while(!stack.empty()) {
+        cout << stack.top();
+        stack.pop();
+    }
+    cout << "\n\n";
+
+    /*************** SETS ***************/
+    // http://www.cplusplus.com/reference/set/set/?kw=set
+    // init
+    // set<TYPE> name
+    set<string> ordered_set {"1", "5", "6"};
+
+    // can insert elements into set
+    ordered_set.insert("2");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    // If the same element is inserted TODO
+    ordered_set.insert("1");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    // elements can also be erased
+    ordered_set.erase("1");
+    print_data(ordered_set.begin(), ordered_set.end(), "ordered_set");
+
+    /*************** UNORDERED SETS ***************/
+    // http://www.cplusplus.com/reference/unordered_set/unordered_set/
+    // init
+    // set<TYPE> name
+    unordered_set<string> unordered_set {"20","19","18","1","16"};
+
+    // can insert elements into set
+    unordered_set.insert("21");
+    // order is not maintained here
+    print_data(unordered_set.begin(), unordered_set.end(), "unordered_set");
+
+    /*************** MAPS ***************/
```

Comment:
The maps don't show off the unordered vs plain distinction, since we only add one element. Not sure if we should add it though.